photon-api/src/main/scala/com/linkedin/photon/ml/data/LocalDataSet.scala

```diff
@@ -212,6 +254,116 @@ object LocalDataSet {
+   *
+   * @param randomLabelAndFeatures An [[Array]] of (label, feature vector) tuples
+   * @param zScore The quartile quantity of standard normal distribution
+   * @param globalFeatureShardStats The global population feature statistics to help compute ratio
+   * @return the lowerBounds for feature columns
+   */
+  protected[ml] def computeRatioCILowerBound(
+      randomLabelAndFeatures: Array[(Double, Vector[Double])],
+      zScore: Double,
+      globalFeatureShardStats: BasicStatisticalSummary,
+      globalPositiveInstances: Array[Double],
+      binaryIndices: Set[Int],
+      epsilon: Double = 0.5): Map[Int, Double] = {
+
+    val lowerBounds = mutable.Map[Int, Double]()
+    val n = globalFeatureShardStats.numNonzeros.toArray
+    val y = globalPositiveInstances
+    val numFeatures = randomLabelAndFeatures(0)._2.size
+
+    val m = randomLabelAndFeatures.map(_._2).toSeq.reduce(_ + _)
+    val x = randomLabelAndFeatures.filter(_._1 == 1).map(_._2).toSeq.reduceLeftOption(_ + _).getOrElse(Vector.zeros[Double](numFeatures))
+
+    binaryIndices.foreach { key => {
+
+        // Do computation on only binary and non-intercept features
+        val m_col = m(key)
+        val n_col = n(key)
+        val x_col = x(key)
+        val y_col = y(key)
+
+        val lowerBound =
+          if (x_col == 0.0 && y_col == 0.0) {
+            0D
+          } else if (x_col == 0.0 && y_col != 0.0) {
+            val (t, variance) = computeTAndVariance(epsilon, m_col, y_col, n_col)
+
+            1D / computeUpperBound(t, variance, zScore)
+          } else if (x_col == m_col && y_col == n_col) {
+            val (t, variance) = computeTAndVariance(m_col - epsilon, m_col, n_col - epsilon, n_col)
+            if (t < 1D) {
+              1D / computeUpperBound(t, variance, zScore)
+            } else {
+              computeLowerBound(t, variance, zScore)
+            }
+          } else {
+            val (t, variance) = computeTAndVariance(x_col, m_col, y_col, n_col)
+
+            if (t < 1D) {
+              1D / computeUpperBound(t, variance, zScore)
+            } else {
+              computeLowerBound(t, variance, zScore)
+            }
+          }
+        lowerBounds.update(key, lowerBound)
+      }
+    }
+
+    lowerBounds
+  }
+
+  /**
+   * Compute t value and variance for ratio modelling.
+   *
+   * @param x The count for f_i == 1 and label == 1 in the local population
+   * @param m The count for f_i == 1 in the local population
+   * @param y The count for f_i == 1 and label == 1 in the global population
+   * @param n The count for f_i == 1 in the global population
+   * @return The mean and variance for ratio t
+   */
+  protected[ml] def computeTAndVariance(
+      x: Double,
+      m: Double,
+      y: Double,
+      n: Double): (Double, Double) = {
+
+    if(m == 0 || n == 0)
+      return (0.0, 0.0)
+    val t = (x / m) / (y / n)
+    val variance = 1.0 / x - 1.0 / m + 1.0 / y - 1.0 / n
+    (t, variance)
```

Comment:
Using `return` for early exit / control flow like this is generally frowned upon in the Scala world. A more idiomatic way would be:

```scala
  if (m == 0 || n == 0) {
    (0.0, 0.0)

  } else {
    val t = (x / m) / (y / n)
    val variance = 1.0 / x - 1.0 / m + 1.0 / y - 1.0 / n

    (t, variance)
  }
```

Reply:
fixed